internal/datastore/proxy/caching.go

```diff
@@ -78,7 +78,9 @@ func (r *nsCachingReader) ReadNamespace(
 		// We couldn't use the cached entry, load one
 		var err error
 		loadedRaw, err, _ = r.p.readNsGroup.Do(nsRevisionKey, func() (any, error) {
-			loaded, updatedRev, err := r.Reader.ReadNamespace(ctx, nsName)
+			// sever the context so that another branch doesn't cancel the
+			// single-flighted namespace read
+			loaded, updatedRev, err := r.Reader.ReadNamespace(context.Background(), nsName)
```

Comment:
> What happens if this request takes unexpectedly long to return?

The grpc stack is still in control of timeouts, so the request will get cancelled (it just won't cancel the ReadNamespace call too).

> Would that affect latencies of callers of this method? 

Singleflight collapses all concurrent requests into one call, so a slow ReadNamespace will definitely slow down any calls waiting for the result, but it won't build up requests for the same namespace/revision and muck things up further.

If ReadNamespaces is slow and requests move forward (so they're requesting newer revisions), then having a bunch of old, unwanted ReadNamespace calls still running does make things worse. But it's almost certainly better for that to potentially happen than to not use singleflight for ReadNamespace.

> Also what happens with tracing, wouldn't we loose it?

Good point, I've done the same thing we do in the datastore and persisted the span/logger/etc from the context. Only one of them will "win" though.

Reply:
Do we think that it's worth implementing our own singleflight that separates the context for us? That way we can avoid having this issue in the future.

Reply:
I did do that just to see how it would look: https://github.com/ecordell/spicedb/pull/1

IMO it's not really any better, since the singleflight API takes a function and it can always close over any context that's not the one passed in the arg, it doesn't _really_ prevent this.

Reply:
>If ReadNamespaces is slow and requests move forward (so they're requesting newer revisions), then having a bunch of old, unwanted ReadNamespace calls still running does make things worse. But it's almost certainly better for that to potentially happen than to not use singleflight for ReadNamespace.

I guess I'm wondering how could we keep track of all the requests that are waiting for a specific key, and cancel the single flight only when all waiting requests context have been cancelled. Does single flight not offer a means to determine when there are no more requests waiting?

Reply:
> I guess I'm wondering how could we keep track of all the requests that are waiting for a specific key, and cancel the single flight only when all waiting requests context have been cancelled. Does single flight not offer a means to determine when there are no more requests waiting?

There's nothing there to help with that - note that singlflight just takes a function as an arg, we just happen to be closing over a context.Context. 

Jimmy asked about something similar, and I did spend some time thinking about it, but came away not thinking there was a good solution that still included singleflight (we'd need to work out our own implementation and interface). Here are my notes on that:

- Ideally the context that actually runs the singleflighted request would get cancelled if all parent contexts are cancelled
- But that means tracking parent contexts when they're added
- Knowing when parent contexts are cancelled means reading from their done channels (or polling their `Err()`s), which is either a goroutine per request (blech) or reflect.Select (blech) or a busy loop (also blech).
- Adding/removing parent context watches probably(?) needs to be tied to the singleflight mutex because we need to know if the request is in-flight or not to know whether to track the context, so context tracking this way will slow down the actual request resolution
- Even when adding a context to the singleflight interface, nothing stops you from using another context from the closure. so the protection would be minimal at best